src/ripple/basics/impl/Number.cpp

```diff
@@ -0,0 +1,729 @@
+            --xm;
+            if (xm < minMantissa)
+            {
+                xm *= 10;
+                --xe;
+            }
+        }
+        if (xe < minExponent)
+        {
+            xm = 0;
+            xe = Number{}.exponent_;
+        }
+    }
+    mantissa_ = xm * xn;
+    exponent_ = xe;
+    return *this;
+}
+
+Number&
+Number::operator*=(Number const& y)
+{
+    if (*this == Number{})
+        return *this;
+    if (y == Number{})
+    {
+        *this = y;
+        return *this;
+    }
+    assert(isnormal() && y.isnormal());
+    auto xm = mantissa();
+    auto xe = exponent();
+    int xn = 1;
+    if (xm < 0)
+    {
+        xm = -xm;
+        xn = -1;
+    }
+    auto ym = y.mantissa();
+    auto ye = y.exponent();
+    int yn = 1;
+    if (ym < 0)
+    {
+        ym = -ym;
+        yn = -1;
+    }
+    auto zm = uint128_t(xm) * uint128_t(ym);
+    auto ze = xe + ye;
+    auto zn = xn * yn;
+    Guard g;
+    while (zm > maxMantissa)
+    {
+        g.push(static_cast<unsigned>(zm % 10));
+        zm /= 10;
+        ++ze;
+    }
+    xm = static_cast<rep>(zm);
+    xe = ze;
+    auto r = g.round();
+    if (r == 1 || (r == 0 && (xm & 1) == 1))
+    {
+        ++xm;
+        if (xm > maxMantissa)
+        {
+            xm /= 10;
+            ++xe;
+        }
+    }
+    if (xe < minExponent)
+    {
+        xm = 0;
+        xe = Number{}.exponent_;
+    }
+    if (xe > maxExponent)
+        throw std::overflow_error(
+            "Number::multiplication overflow : exponent is " +
+            std::to_string(xe));
+    mantissa_ = xm * zn;
+    exponent_ = xe;
+    assert(isnormal() || *this == Number{});
+    return *this;
+}
+
+Number&
+Number::operator/=(Number const& y)
+{
+    if (y == Number{})
+        throw std::overflow_error("Number: divide by 0");
+    int np = 1;
+    auto nm = mantissa();
+    if (nm < 0)
+    {
+        nm = -nm;
+        np = -1;
+    }
+    int dp = 1;
+    auto dm = y.mantissa();
+    if (dm < 0)
+    {
+        dm = -dm;
+        dp = -1;
+    }
+    // Divide numerator and denominator such that the
+    //     denominator is in the range [1, 10).
+    const int offset = -15 - y.exponent();
+    Number n{nm * (np * dp), exponent() + offset};
+    Number d{dm, y.exponent() + offset};
+    // Quadratic least squares fit to 1/x in the range [1, 10]
+    constexpr Number a0{9178756872006464, -16, unchecked{}};
+    constexpr Number a1{-2149215784206187, -16, unchecked{}};
+    constexpr Number a2{1405502114116773, -17, unchecked{}};
+    static_assert(a0.isnormal());
+    static_assert(a1.isnormal());
+    static_assert(a2.isnormal());
+    Number rm2{};
+    Number rm1{};
+    Number r = (a2 * d + a1) * d + a0;
+    //  Newton–Raphson iteration of 1/x - d with initial guess r
+    //  halt when r stops changing, checking for bouncing on the last iteration
+    do
+    {
+        rm2 = rm1;
+        rm1 = r;
+        r = r + r * (one - d * r);
+    } while (r != rm1 && r != rm2);
```

Comment:
Yes, both by theory and testing.

Theory:  The Newton algorithm will reliably point towards the right answer.  If that right answer is between two representable values, and the starting guess is one of those representable values, then the algorithm either gives back the same answer, or the other representable value closest to the exact answer.  It never "wanders" away.

Practice:  The set up to this look reduces the domain to [1, 10).  Very significant testing was done throughout this small finite range, especially near the endpoints, and at the point where the difference between the initial guess and the exact answer is at a maximum.

I also tried going with a fixed number of iterations.  However testing revealed that the required number of iterations varied enough that I felt it was too expensive to hard code the maximum number of iterations when many inputs in the range [1, 10) required fewer than the maximum.

In the end, among these four approaches:

1.  Hard code the number of iterations:  Rejected because there was no one number of iterations that was acceptable.
2. Stop when r == rm1.  Rejected because the loop sometimes did not terminate because the algorithm would bounce between two values.
3. Stop when r == rm1 || r == rm2.  Accepted as the highest performance alternative while maintaining correctness.
4. Stop when r == rm1 || r == rm2 || r == rm3.  Although correct, a history of three values was never observed to be needed.  Rejected.